apps/els_lsp/src/els_db_server.erl

```diff
@@ -83,7 +93,24 @@ handle_call({match_delete, Table, Pattern}, _From, State) ->
   {reply, ok, State};
 handle_call({write, Table, Object}, _From, State) ->
   true = ets:insert(Table, Object),
-  {reply, ok, State}.
+  {reply, ok, State};
+handle_call({conditional_write, Table, Key, Object, Condition}, _From, State) ->
+  case ets:lookup(Table, Key) of
+    [Entry] ->
+      case Condition(Entry) of
+        true ->
+          true = ets:insert(Table, Object),
+          {reply, ok, State};
+        false ->
+          ?LOG_DEBUG("Skip insertion due to invalid condition "
+                     "[table=~p] [key=~p]",
+                     [Table, Key]),
+          {reply, {error, condition_not_satisfied}, State}
+      end;
+    [] ->
+      true = ets:insert(Table, Object),
+      {reply, ok, State}
+  end.
```

Comment:
I wonder, if this could use ETS's `select_replace` instead for atomic writes directly in ETS. This still has a small chance of race conditions between when you look up the value and write it - the underlying database could have changed.

Reply:
Considering all writes to the ETS are serialized via a gen_server, this should not be a problem in practice.